### PR TITLE
Icon cache and Task.Run tweaks

### PR DESCRIPTION
**Pull request by hez2010**
**Details of Changes**
Add details of changes here.
- Introduced a cache for icon and overlay
- Avoid using `Task.Run` for long-running tasks

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [x] Tested the changes for accessibility

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

File: src/Files.App/ViewModels/ItemViewModel.cs [hunk 1]
```diff
@@ -1878,7 +1878,7 @@ private void WatchForDirectoryChanges(string path, CloudDriveSyncStatus syncStat
 
             if (aProcessQueueAction == null) // Only start one ProcessOperationQueue
             {
-                aProcessQueueAction = Task.Run(() => ProcessOperationQueue(watcherCTS.Token, hasSyncStatus));
+                aProcessQueueAction = Win32API.StartSTATask(() => ProcessOperationQueue(watcherCTS.Token, hasSyncStatus));
```
Comment (gave92) [inline src/Files.App/ViewModels/ItemViewModel.cs:1881]: Not a problem but my understanding is that a dedicated thread makes sense only if you have blocking & non cpu-bound work (most of the work here is short running or asynchronous).

If instead there's a need to specify the task is long running perhaps I'd use Task.Factory.StartNew + LongRunning option, I'd reserve StarSTATask for operations that really require a STA thread + OLE (context menu, shell thumbnails..)

Comment (hez2010) [reply src/Files.App/ViewModels/ItemViewModel.cs]: Done.

Comment (gave92) [reply src/Files.App/ViewModels/ItemViewModel.cs]: Thanks, unfortunately Task.factory.StartNew has some bad default parameters, I believe the correct call to be smth like (note expecially the TaskScheduler):

`Task.Factory.StartNew(someAction, CancellationToken.None, TaskCreationOptions.DenyChildAttach | TaskCreationOptions.LongRunning, TaskScheduler.Default)`

Comment (gave92) [reply src/Files.App/ViewModels/ItemViewModel.cs]: Note: my knowledge about this topic is a bit poor, so feel free to disregard these suggestions / wskt for other oponions ;)

File: src/Files.App/Shell/Win32API.cs [hunk 2]
```diff
@@ -16,13 +19,75 @@
...
+                Priority = ThreadPriority.Normal
+            };
+            thread.SetApartmentState(ApartmentState.STA);
+            thread.Start();
+            return tcs.Task;
+        }
+
+        public static Task StartSTATask(Action action)
+        {
+            var tcs = new TaskCompletionSource();
```
Comment (yaichenbaum) [inline src/Files.App/Shell/Win32API.cs:63]:
```suggestion
            var taskCompletionSource = new TaskCompletionSource();
```
Can you do the same for the other variables? It helps make the code more readable.

File: src/Files.App/Shell/Win32API.cs [hunk 3]
```diff
@@ -319,37 +426,37 @@ public static IList<IconFileInfo> ExtractSelectedIconsFromDLL(string file, IList
...
             if (icoCnt == null)
             {
                 return null;
             }
+
             int count = icoCnt.DangerousGetHandle().ToInt32();
-            int maxIndex = count - 1;
-            if (maxIndex == 0)
+            if (count <= 0)
             {
```
Comment (yaichenbaum) [inline src/Files.App/Shell/Win32API.cs:438]: The brackets can be removed for single line if statements.

File: src/Files.App/Shell/Win32API.cs [hunk 4]
```diff
@@ -362,7 +469,34 @@ public static bool SetCustomDirectoryIcon(string folderPath, string iconFile, in
...
+                _iconAndOverlayCache[folderPath] = new();
+            }
+            return success;
+        }
+
+        public static async Task<bool> SetCustomFileIconAsync(string filePath, string iconFile, int iconIndex = 0)
+        {
+            var connection = await AppServiceConnectionHelper.Instance;
+            if (connection != null)
+            {
```
Comment (yaichenbaum) [inline src/Files.App/Shell/Win32API.cs:484]:
```suggestion
            if (connection == null)
                        return;
```
Returning early can help make the code more readable 👍

**Review by yaichenbaum — APPROVED** [review]

File: src/Files.App/ViewModels/ItemViewModel.cs (hunk 1 above)
Comment (hez2010) [reply src/Files.App/ViewModels/ItemViewModel.cs]: I just checked the document and did't see why we need `TaskCreationOptions.DenyChildAttach`.

File: src/Files.App/Shell/Win32API.cs [hunk 5]
```diff
@@ -428,16 +428,10 @@ public static IList<IconFileInfo> ExtractIconsFromDLL(string file)
             var iconsList = new List<IconFileInfo>();
             using var currentProc = Process.GetCurrentProcess();
             using var icoCnt = Shell32.ExtractIcon(currentProc.Handle, file, -1);
-            if (icoCnt == null)
-            {
-                return null;
-            }
+            if (icoCnt == null) return null;
```
Comment (yaichenbaum) [inline src/Files.App/Shell/Win32API.cs:431]: I think `return` should still be on the next line, just without the brackets.

Comment (hez2010) [reply src/Files.App/Shell/Win32API.cs]: Done.

**Review by yaichenbaum — APPROVED** [review]
LGTM

File: src/Files.App/ViewModels/ItemViewModel.cs (hunk 1 above)
Comment (gave92) [reply src/Files.App/ViewModels/ItemViewModel.cs]: It would have been only to make it equivalent to Task.Run, but not necessary.

Comment (gave92) [reply src/Files.App/ViewModels/ItemViewModel.cs]: PS the important point was TaskScheduler.Default btw. Not passing it risks running the Task on the ui thread / causing a deadlock.

Comment (hez2010) [reply src/Files.App/ViewModels/ItemViewModel.cs]: Will fix this in following PR.

**Pull request merged**